src/main.rs

```diff
@@ -37,7 +39,9 @@ async fn run() -> Result<Infallible, HopperError> {
         .unwrap_or_else(|| Box::new(EmptyInjector));
 
     // builds a new hopper instance with a router
-    let server = Hopper::new(Arc::new(config.routing), metrics);
+    let server = Hopper::new(Arc::new(Mutex::new(Arc::new(config.routing))), metrics);
+
+    server.listen_config(signal(SignalKind::hangup()).expect("Unable to get signal"));
```

Comment:
Instead of spawning a new task you could select the future on L46 with the stop signal wait.